crates/accelerate/src/circuit_library/multi_local.rs

```diff
@@ -0,0 +1,317 @@
+/// - `parameter_prefix`: The prefix used if default parameters are generated.
+/// - `skip_final_rotation_layer`: Whether a final rotation layer is added to the circuit.
+/// - `skip_unentangled_qubits`: If ``True``, the rotation gates act only on qubits that
+///     are entangled. If ``False``, the rotation gates act on all qubits.
+///
+/// # Returns
+///
+/// An N-local circuit.
+#[allow(clippy::too_many_arguments)]
+pub fn n_local(
+    py: Python,
+    num_qubits: u32,
+    rotation_blocks: &[&Block],
+    entanglement_blocks: &[&Block],
+    entanglement: &Entanglement,
+    reps: usize,
+    insert_barriers: bool,
+    parameter_prefix: &String,
+    skip_final_rotation_layer: bool,
+    skip_unentangled_qubits: bool,
+) -> PyResult<CircuitData> {
+    // Construct the parameter ledger, which will define all free parameters and provide
+    // access to them, given an index for a layer and the current gate to implement.
+    let ledger = ParameterLedger::from_nlocal(
+        py,
+        num_qubits,
+        reps,
+        entanglement,
+        rotation_blocks,
+        entanglement_blocks,
+        skip_final_rotation_layer,
+        parameter_prefix,
+    )?;
+
+    // Compute the qubits that are skipped in the rotation layer. If this is set,
+    // we skip qubits that do not appear in any of the entanglement layers.
+    let skipped_qubits = if skip_unentangled_qubits {
+        let active: HashSet<&u32> =
+            HashSet::from_iter(entanglement.iter().flatten().flatten().flatten());
+        HashSet::from_iter((0..num_qubits).filter(|i| !active.contains(i)))
+    } else {
+        HashSet::new()
+    };
+
+    // This struct can be used to yield barrier if insert_barriers is true, otherwise
+    // it returns an empty iterator. For conveniently injecting barriers in-between operations.
+    let maybe_barrier = MaybeBarrier::new(py, num_qubits, insert_barriers)?;
+
+    let mut packed_insts: Box<dyn Iterator<Item = PyResult<Instruction>>> =
+        Box::new((0..reps).flat_map(|layer| {
+            rotation_layer(
+                py,
+                num_qubits,
+                rotation_blocks,
+                ledger.get_parameters(LayerType::Rotation, layer),
+                &skipped_qubits,
+            )
+            .chain(maybe_barrier.get())
+            .chain(entanglement_layer(
+                py,
+                entanglement.get_layer(layer),
+                entanglement_blocks,
+                ledger.get_parameters(LayerType::Entangle, layer),
+            ))
+            .chain(maybe_barrier.get())
+        }));
+    if !skip_final_rotation_layer {
+        packed_insts = Box::new(packed_insts.chain(rotation_layer(
+            py,
+            num_qubits,
+            rotation_blocks,
+            ledger.get_parameters(LayerType::Rotation, reps),
+            &skipped_qubits,
+        )))
+    }
+
+    CircuitData::from_packed_operations(py, num_qubits, 0, packed_insts, Param::Float(0.0))
```

Comment:
You don't need a box here, you can do something like:

```suggestion
    let packed_insts = (0..reps).flat_map(|layer| {
        rotation_layer(
            py,
            num_qubits,
            rotation_blocks,
            ledger.get_parameters(LayerType::Rotation, layer),
            &skipped_qubits,
        )
        .chain(maybe_barrier.get())
        .chain(entanglement_layer(
            py,
            entanglement.get_layer(layer),
            entanglement_blocks,
            ledger.get_parameters(LayerType::Entangle, layer),
        ))
        .chain(maybe_barrier.get())
    });
    if !skip_final_rotation_layer {
        let packed_insts = packed_insts.chain(rotation_layer(
            py,
            num_qubits,
            rotation_blocks,
            ledger.get_parameters(LayerType::Rotation, reps),
            &skipped_qubits,
        ));
        CircuitData::from_packed_operations(py, num_qubits, 0, packed_insts, Param::Float(0.0))
    } else {
        CircuitData::from_packed_operations(py, num_qubits, 0, packed_insts, Param::Float(0.0))
    }
```

The trick is to avoid the dynamic type when you call `from_packed_operations`